### ISSUE TITLE
Upload multiple public SSH keys to DbSystem

### ISSUE THREAD

**Issue by damtst** [terraform-providers/terraform-provider-oci#1007]
Hello, I am trying to upload several public ssh keys to a dbsystem using terraform. Whilst this works for instances, on dbsystems everytime I run the terraform apply, it replaces the keys.

I have tried from a file, a string with newline separation and multi-line string with the same results:
`
      ~ ssh_public_keys         = [
          + <<~EOT
                ssh-rsa kUYVBzMioKCtbazK5U+lkDS0NBmN5sbzIpE09zTZ8nRswHuDy9XlmFI3ILwGezeyVy9jbNUjdSXhXioVpo+l+ZGFRExWKiDUkUkCWeFew19SmlWAvqpI5WY8bDkeSXNz6u2DhzziJ10fKdKWD4dsDDOGO4lkxKf6fa5eWl4t5PLeEKOpleD3wu9Qy4rSjXZHhk4qXnxzhLvFMm5Sx4SR7jQ48jGM3eGzX5aG4/Ovbq3JQ2wimoZDIeWOt//zdl/QVCIflpbFt99Xa+WKnPFuen2pjF4ddN9xQ7uDSZxDHKz9Ebqx8AWU6C/PsE+2ilfktWO5I4xUX ppp@yyy
                ssh-rsa G3Q4bHVEd49DjkKaN/l/iMnE/7OkD8XilihADF/XOov32deX8hTA3EdFc5C4sjlSDk1NqbT16ADK2tf7//9TpzndmSs3yNQyTIyJSmDE+3baPONP6Z/bMkdrQdB9xeICyO5QxeB3/IF/uU1ZBCmGVd++ch/cvqcLC72VjqFEtF738PpbAr0ryID+EaTmKU0BMCTUXfGvEHgcqBN/GJp3ooUMJPKQURbl6AKG1wIqrarxw3a0V25RUDB/f6HiXzpVHEkED76XViRYOSW795fSIJnkejwnMzYQCL2hcmPMXmve1kId ddd@yyy
                ssh-rsa +dZGnPfn0Hs60aaL0FhSxu2nnCGDZuHxN1mhv4aLxElo2CZo0j5bpNHduJsJGo4DMV3uMg94sW5ebVEZuLfu+EL70J9c8K2/vvlW++6whPB9SrYw3Czd/77TMJ5/eMk1SlcTMy7d70hA4wOtBMsFANcV3edrYi1v3c5Kw8m7XAT2jIDWpV5D56BqruSLk6bGoKzu+dniXN/VvL3VGBt89KMV4r3W6Q25X/praCB663Ezy4enYY7oFM53VS0n/05ilhynyq//YAw03oh2RlAhZxgJQfQK8y7zGf7/6+bhLt8vhgY8I/VKmiwsU1 zzz@yyy
            EOT,
          - "ssh-rsa kUYVBzMioKCtbazK5U+lkDS0NBmN5sbzIpE09zTZ8nRswHuDy9XlmFI3ILwGezeyVy9jbNUjdSXhXioVpo+l+ZGFRExWKiDUkUkCWeFew19SmlWAvqpI5WY8bDkeSXNz6u2DhzziJ10fKdKWD4dsDDOGO4lkxKf6fa5eWl4t5PLeEKOpleD3wu9Qy4rSjXZHhk4qXnxzhLvFMm5Sx4SR7jQ48jGM3eGzX5aG4/Ovbq3JQ2wimoZDIeWOt//zdl/QVCIflpbFt99Xa+WKnPFuen2pjF4ddN9xQ7uDSZxDHKz9Ebqx8AWU6C/PsE+2ilfktWO5I4xUX ppp@yyy ssh-rsa G3Q4bHVEd49DjkKaN/l/iMnE/7OkD8XilihADF/XOov32deX8hTA3EdFc5C4sjlSDk1NqbT16ADK2tf7//9TpzndmSs3yNQyTIyJSmDE+3baPONP6Z/bMkdrQdB9xeICyO5QxeB3/IF/uU1ZBCmGVd++ch/cvqcLC72VjqFEtF738PpbAr0ryID+EaTmKU0BMCTUXfGvEHgcqBN/GJp3ooUMJPKQURbl6AKG1wIqrarxw3a0V25RUDB/f6HiXzpVHEkED76XViRYOSW795fSIJnkejwnMzYQCL2hcmPMXmve1kId ddd@yyyssh-rsa +dZGnPfn0Hs60aaL0FhSxu2nnCGDZuHxN1mhv4aLxElo2CZo0j5bpNHduJsJGo4DMV3uMg94sW5ebVEZuLfu+EL70J9c8K2/vvlW++6whPB9SrYw3Czd/77TMJ5/eMk1SlcTMy7d70hA4wOtBMsFANcV3edrYi1v3c5Kw8m7XAT2jIDWpV5D56BqruSLk6bGoKzu+dniXN/VvL3VGBt89KMV4r3W6Q25X/praCB663Ezy4enYY7oFM53VS0n/05ilhynyq//YAw03oh2RlAhZxgJQfQK8y7zGf7/6+bhLt8vhgY8I/VKmiwsU1 zzz@yyy",
        ]
`

this is the version we are using:
`
Terraform v0.12.20
+ provider.oci v3.66.0
`

Thank you.

**Comment by damtst** [comment]
Hello, 

just uploaded to the latest terraform version and we have the same results:
`Terraform v0.12.24
+ provider.oci v3.65.0
+ provider.template v2.1.2`

thank you

**Comment by damtst** [comment]
Hello, 
tried with OCI v3.69 with the same result.

can someone please let us know how to upload multiple public ssh keys to a dbsystem without having to re-do the same operation with every terraform apply?

thank you.

**Comment by alexng-canuck** [comment]
This seems like a case where Terraform found a mismatch between your config's keys and the keys that we were able to retrieve from the DB service.

This could be because Terraform thinks the `<<EOT` syntax is part of your expected keys and causing a diff; when in fact the actual keys returned by the service don't have the `<<EOT` syntax.

It could also be because the returned set of SSH keys from the service has a slight format difference (note the space difference between `ddd@yyy ssh-rsa` in the keys being added vs. `ddd@yyyssh-rsa` in the keys being removed)

Can you share your configs? Instead of using Heredoc syntax, have you tried specifying multiple separate SSH keys like this?

```
ssh_public_keys = [ "ssh key 1", "ssh key 2",  "ssh key 3"]
```

Since the `ssh_public_keys` field is supposed to be a list of strings, the keys should be passed in separately as a Terraform list.

**Comment by damtst** [comment]
Hello Alex, 

first of all, thank you for your help, much appreciated.

This is the three configurations I have tested so far:
1. heredoc:
`ssh_public_key =<<-EOT
                   ssh-rsa AAAAB34xUX xxx@ccc 
                   ssh-rsa PMXmve1kId yyy@bbb
                   ssh-rsa rGVKmiwsU1 zzz@aaa
                EOT
`

2. variable (all key with return of line separation):
`
ssh-rsa AAAAB34xUX xxx@ccc\nssh-rsa PMXmve1kId yyy@bbb\nssh-rsa rGVKmiwsU1 zzz@aaa\n
`
3. file containing the public ssh (one key per line:
`
ssh-rsa AAAAB34xUX xxx@ccc 
ssh-rsa PMXmve1kId yyy@bbb
ssh-rsa rGVKmiwsU1 zzz@aaa
`

but in all cases it replaces the ssh key on the server as per initial post. All configurations above works fine with instances so I am assuming is the difference of type.

can you please let me know if it will be possible to have the same type for ssh_public_key for dbsystem and for instances?

I will also try your solution in the next hour but it looks very good.

thank you

David

**Comment by damtst** [comment]
Hello, 

the solution of having ssh keys as ["xxx","yyy","zzz"] is good and it works.

we have now an issue with the following points, bearing in mind we have the keys on a file:
1. if we remove a ssh key from the file, terraform will modify the dbsystem or recreate the instance (can we avoid the recreation of a vm?).
2. if we remove the ssh key from the dbsystem, terraform does not see the difference hence do nothing. I do not think this is correct, I think Terraform should add the missing key.
3. if we have spaces at the end of the ssh key, terraform thinks it is a different key so replaces it, even when we are using trimspace.Checking the key on the server, it is uploaded without spaces at the end.

below is a test I have created:

id_rsa_group.pub:
`
ssh-rsa AAAABAQC3E6QWwTPPH xxx@aaa 
ssh-rsa AAAABAQDzw2t3zbTBQ yyy@bbb     
ssh-rsa AAAABAQDDomlZgj6nU zzz@ccc
`
notice there are 1 space after the first key, 4 after the second and none after the third.

as you will see below, we read the file (file), remove spaces (trimspace), create a new string after newline (split) and create a list (tolist)

main.tf:
`
variable "tenancy_ocid" {}
variable "user_ocid" {}
variable "fingerprint" {}
variable "private_key_path" {}
variable "compartment_ocid" {}
variable "region" {}

variable "ssh_public_key" {}

provider "oci" {
version = ">=3.69"
tenancy_ocid = var.tenancy_ocid
user_ocid = var.user_ocid
fingerprint = var.fingerprint
private_key_path = var.private_key_path
region = "xxxx"
alias = "yyy"
disable_auto_retries = "true"
}


locals {
  db_ssh_public_key = file("${path.module}/id_rsa_group.pub")
  db_ssh_public_key2 = "${path.module}/id_rsa_group.pub"
}

output "a" {
value = local.db_ssh_public_key
}

output "b" {
value = formatlist ("%s",local.db_ssh_public_key)
}

output "c" {
value = split ("\r\n",local.db_ssh_public_key)
}

output "d" {
value = tolist(split ("\r\n",local.db_ssh_public_key))
}

output "e" {
  value = split ("\r\n",trimspace (file(local.db_ssh_public_key2)))
}

`
on the different outputs, you will see that spaces at the end are not removed.

Thank you.

**Comment by varmax2511** [comment]
Hi @damtst 

Responding to your questions:
1. When removing SSH key from file, terraform will recognize a change in the field and suggest an in-place update unless you are also modifying some other parameter which is a `ForceNew`. Field `ssh_keys_public` is updatable and will not recreate the dbsystem.
Please refer: https://www.terraform.io/docs/providers/oci/r/database_db_system.html#ssh_public_keys

2. Can you please explain `remove the key from dbSystem`? If you are removing the field `ssh_public_keys` from the terraform config or updating the key value passed, terraform detects the change

3. Please refer to https://www.terraform.io/docs/configuration/functions/trimspace.html. In your scenario you are applying trimspace before splitting the list items and hence
it trims the space of the overrall string and not the individual lines defined within in the string
```
ssh-rsa AAAABAQC3E6QWwTPPH xxx@aaa 
ssh-rsa AAAABAQDzw2t3zbTBQ yyy@bbb     
ssh-rsa AAAABAQDDomlZgj6nU zzz@ccc
```

Please let us know if this helps?

**Comment by varmax2511** [comment]
Hi @damtst -- can you please let us know if the issue is resolved?

**Comment by nguyenhv75** [comment]
Yes the issue has been resolved.
The code was from 2017 and need to be updated with current settings.
Thank you for looking into this.

**Comment by varmax2511** [comment]
thanks @nguyenhv75.

**Comment by damtst** [comment]
Hello all, 

sorry for the delay, I did not manage to connect during Easter break.

couple of things:
1. trimspace does not work for us, only removes empty lines.
2. I managed to remove trailing spaces by using replace 
    `replace(file("${path.module}/id_rsa_group.pub"),"/ *\r\n/","\r\n"))`
   so my expression to retrieve public keys from a file and remove trailing spaces and empty lines is
`tolist(split ("\r\n",trimspace(replace(file("${path.module}/id_rsa_group.pub"),"/ *\r\n/","\r\n"))))`
3. if we connect to a dbsystem using opc and remove one key from .ssh/authorized_keys, terraform terraform does not recognize the missing key and does not re-import them.
4. it is true that on a dbsystem, changing the keys will be an update of the resource, but on a instance (vm for example) it will trigger a forcenew. I think instances should not be destroy and created by adding a new public key, please find an example below
  # module.create_spoke3.module.create_bastion.oci_core_instance.test_instance[0] must be replaced
```
-/+ resource "oci_core_instance" "test_instance" {
      ~ launch_mode                         = "NATIVE" -> (known after apply)
      ~ metadata                            = { # forces replacement
          ~ "ssh_authorized_keys" = <<~EOT
              - ssh-rsa AAAAB3NzaC1ycO5I4xUX xxx@yyy
              + ssh-rsa AAAAB3NzaC1ycO5I4xUX xxx@yyy
                ssh-rsa AAAAB3NzaC1ycmve1kId zzz@fff
                ssh-rsa AAAAB3NzaC1ycKmiwsU1 rrr@uuu
            EOT
            "user_data"           = "IyEvYmluL2Jhc2gNCnl1bSB1cGRhdGUgLXkNCnN1ZG8geXVtIGluc3RhbGwgdGNwaW5nIC15DQo="
        }
```

@nguyenhv75, can you please let me know what issue has been fixed due to code being from 2017?

thank you, again, for all your help.

David.

**Comment by nguyenhv75** [comment]
Brackets should not be use for the following:

   - availability_domain =
   ["${data.template_file.deployment_ad.*.rendered}"]  + availability_domain =
   "${element(var.availability_domain, count.index)}" == Failed

By removing brackets, the command using element and count.index will work:

   - availability_domain =
   "${data.template_file.deployment_ad.*.rendered}"  + availability_domain =
   "${element(var.availability_domain, count.index)}" == Succeeded

Another problem is that following image has been pulled by Oracle Support:

   - "${data.oci_core_images.InstanceImageOCID.images.0.id}"

Terraform for OCI best practices for Oracle-Provided Images is show below:

# Get Oracle-Provided Images for Given Region
# OCI Terraform Best Practices -
https://www.terraform.io/docs/providers/oci/guides/best_practices.html
variable "image_id" {
  default = {
    // See https://docs.cloud.oracle.com/iaas/images/
    // Oracle-provided image "Oracle-Linux-7.5-2018.10.16-0"
    us-phoenix-1 =
"ocid1.image.oc1.phx.aaaaaaaaoqj42sokaoh42l76wsyhn3k2beuntrh5maj3gmgmzeyr55zzrwwa"
    us-ashburn-1 =
"ocid1.image.oc1.iad.aaaaaaaageeenzyuxgia726xur4ztaoxbxyjlxogdhreu3ngfj2gji3bayda"
    eu-frankfurt-1 =
"ocid1.image.oc1.eu-frankfurt-1.aaaaaaaaitzn6tdyjer7jl34h2ujz74jwy5nkbukbh55ekp6oyzwrtfa4zma"
    uk-london-1 =
"ocid1.image.oc1.uk-london-1.aaaaaaaa32voyikkkzfxyo4xbdmadc2dmvorfxxgdhpnk6dw64fa3l4jh7wa"
  }
}

Then use following syntax to choose the image:

bastion_image           = "${var.image_id[var.region]}"

**Comment by nguyenhv75** [comment]
Trimspace does work in the following case by enclosed in brackets [] as
follows:

ssh_public_keys         = ["${trimspace(file("${var.db_ssh_public_key}"))}"]

**Comment by damtst** [comment]
Hello, 

I think there is a misunderstanding, `module.create_spoke3.module.create_bastion.oci_core_instance.test_instance[0] must be replaced` 
is the output of executing 
`terraform apply`

I did not highlighted properly, apologies ( i have fixed it).

It basically shows that an instance will be replaced if there is a change on the ssh key whilst a dbsystem will update the keys.

I am also guessing trimspace will not work for this specific parameter (ssh_public_keys), hence my workaround is sound.

any idea of the other points I raised?

thank you, 

David.
